goleveldb.go

```diff
@@ -26,6 +31,12 @@ func NewGoLevelDB(name string, dir string) (*GoLevelDB, error) {
 	return NewGoLevelDBWithOpts(name, dir, nil)
 }
 
+func NewGoLevelDBReadOnly(name string, dir string) (*GoLevelDB, error) {
+	var o opt.Options
+	o.ReadOnly = true
+	return NewGoLevelDBWithOpts(name, dir, &o)
+}
+
```

Comment:
thanks for making a contribution. Could you give some context on why this should be added over passing the option to `NewGoLevelDBWithOpts(..)`?

Reply:
thank u very much for reply. the context:

Our blockchain project (https://github.com/netcloth/netcloth-chain) client needs to be able to sign transactions with high concurrency. There are thousands of local accounts stored in the tm-db with store type goleveldb, but tm-db does not provide APIs that can set goleveldb read-only mode nor does it support setting goleveldb parameters in db instance constructor. The read TPS is about 20 in non-read-only mode by tm-db.

@tessr

Reply:
and The database will not be modified when signing transactions

Reply:
ah I see its because when we call newDB we are essentially prohibiting any DB options to be added. 

what about adding an options arg into here:
`func NewDB(name string, backend BackendType, dir string, opts interface{}) (DB, error) {`

then in each db we check this opts interface to the appropriate DB's options type if its the wrong type we panic or error. this way users cold pass in options. Thoughts? 

cc @erikgrinaker

Reply:
ok， thank u

Reply:
oh sorry before you close this I would like to get eriks thoughts as he has been doing most of the work in this repo.

Reply:
yep, This means that the API interface has been modified, and for the client, all the places that call NewDB need to be modified. A new interface can be added for future clients to use. For the old client code, maybe this PR can help

Reply:
client code:

func NewLevelDB(name, dir string) (db dbm.DB, err error) {
	backend := dbm.GoLevelDBBackend
	if DBBackend == string(dbm.CLevelDBBackend) {
		backend = dbm.CLevelDBBackend
	}

	if strings.Contains(dir, ".nchcli") { // nchcli can readonly for signing txs
		backend = dbm.GoLevelDBBackendReadOnly
	}

	defer func() {
		if r := recover(); r != nil {
			err = fmt.Errorf("couldn't create db: %v", r)
		}
	}()
	return dbm.NewDB(name, backend, dir), err
}

// DBBackend is a module variable

Reply:
We already have backend-specific constructors that allow passing options to the backend database, including read-only support for GoLevelDB:

```go
import (
	tmdb "github.com/tendermint/tm-db"
	"github.com/syndtr/goleveldb/leveldb/opt"
)

db, err := tmdb.NewGoLevelDBWithOpts("gldb", "/tmp", opt.Options{ReadOnly: true})
```

Is this not sufficient?

Reply:
ok, thank u.